### PR TITLE
solver class, memory reallocation, and regularization

### PR DESCRIPTION
**Pull request by Robbybp**
## Summary/Motivation:
Changes to the interior point algorithm.

## Changes proposed in this PR:
- Allow reallocating memory in the linear solver
- Basic regularization strategy
- `InteriorPointSolver` class. My intent is for this to be subclassed for a variety of experimental interior point methods (different options, strategies e.g. for regularization). Not sure exactly what is the best way to organize all the different varieties of the algorithm we want to implement. 

I renamed my old branch (from #3) and created this new one to reflect only changes made to the interior point algorithm, not changes made in order to test it.  

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/interior_point/tests/test_reg.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        # Perform one iteration of interior point algorithm
+        x, duals_eq, duals_ineq = ip_solver.solve(interface, max_iter=1)
+
+# The exact regularization coefficient at which Mumps recognizes the matrix
+# as non-singular appears to be non-deterministic...
+# I have seen 1e-4, 1e-2, and 1e0.
+# According to scipy, 1e-4 seems to be correct
+# MUMPS 5.3.1 seems to settle on 1e-2
+#        # Expected regularization coefficient:
+        self.assertAlmostEqual(ip_solver.reg_coef, 1e-2)
```
Comment (Robbybp) [inline pyomo/contrib/interior_point/tests/test_reg.py]: @carldlaird This is the case where I got self-inconsistent behavior with MUMPS 5.2.1 and inconsistent-with-scipy behavior with 5.3.1

File: pyomo/contrib/interior_point/tests/test_reg.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+# This happens even if I recreate linear_solver and ip_solver.
+# Appears to be non-deterministic
+# Using MUMPS 5.2.1
+# Problem persists with MUMPS 5.3.1
+        # Now perform two iterations of the interior point algorithm.
+        # Because of the way the solve routine is written, updates to the
+        # interface's variables don't happen until the start of the next
+        # next iteration, meaning that the interface has been unaffected
+        # by the single iteration performed above.
+        x, duals_eq, duals_ineq = ip_solver.solve(interface, max_iter=2)
```
Comment (Robbybp) [inline pyomo/contrib/interior_point/tests/test_reg.py]: @carldlaird Got a different (even worse) inconsistency here.

**Comment by Robbybp** [comment]
@michaelbynum Looks good after I merged your branch

**Comment by Robbybp** [comment]
@michaelbynum This should be good to merge. I had to remove the status check in the IP solver reallocation method because `get_infog` is not defined for solvers other than MUMPS, but that should be easily superceded by a `LinearSolverResults` object. Also substituted the regularization test problem for a synthetic one.

**Review by michaelbynum — APPROVED** [review]
@Robbybp Looks great. Thanks for all of this!

**Pull request merged**